src/compile_cache.cc

```diff
@@ -223,13 +229,109 @@ void CompileCacheHandler::ReadCacheFile(CompileCacheEntry* entry) {
+    normalised_string.erase(0, file_scheme.size());
+  }
+
+#ifdef _WIN32
+  if (normalised_string.size() > 2 &&
+      IsWindowsDeviceRoot(normalised_string[0]) &&
+      normalised_string[1] == ':' &&
+      (normalised_string[2] == '/' || normalised_string[2] == '\\')) {
+    normalised_string[0] = ToLower(normalised_string[0]);
+  }
+#endif
+  for (char& c : normalised_string) {
+    if (c == '\\') {
+      c = '/';
+    }
+  }
+
+  if (!normalised_string.empty() && normalised_string.back() == '/') {
+    normalised_string.pop_back();
+  }
+
+  normalised_string = NormalizeString(normalised_string, false, "/");
+  return normalised_string;
+}
+
+// Check if a path looks like an absolute path or file URL.
+static bool IsAbsoluteFilePath(std::string_view path) {
+  if (path.rfind("file://", 0) == 0) {
+    return true;
+  }
+#ifdef _WIN32
+  if (path.size() > 2 && IsWindowsDeviceRoot(path[0]) &&
+      (path[1] == ':' && (path[2] == '/' || path[2] == '\\')))
+    return true;
+  if (path.size() > 1 && path[0] == '\\' && path[1] == '\\') return true;
+#else
+  if (path.size() > 0 && path[0] == '/') return true;
+#endif
+  return false;
+}
+
+static std::string GetRelativePath(std::string_view path,
+                                   std::string_view base) {
+// On Windows, the native encoding is UTF-16, so we need to convert
+// the paths to wide strings before using std::filesystem::path.
+// On other platforms, std::filesystem::path can handle UTF-8 directly.
+#ifdef _WIN32
+  std::wstring wpath = ConvertToWideString(std::string(path), CP_UTF8);
+  std::wstring wbase = ConvertToWideString(std::string(base), CP_UTF8);
+  std::filesystem::path relative =
+      std::filesystem::path(wpath).lexically_relative(
+          std::filesystem::path(wbase));
+  if (relative.empty()) {
+    return std::string();
+  }
+  std::string relative_path = ConvertWideToUTF8(relative.wstring());
+  return relative_path;
+#else
+  std::filesystem::path relative =
+      std::filesystem::path(path).lexically_relative(
+          std::filesystem::path(base));
+  if (relative.empty()) {
+    return std::string();
+  }
+  return relative.generic_string();
+#endif
+}
+
 CompileCacheEntry* CompileCacheHandler::GetOrInsert(Local<String> code,
                                                     Local<String> filename,
                                                     CachedCodeType type) {
   DCHECK(!compile_cache_dir_.empty());
 
   Utf8Value filename_utf8(isolate_, filename);
-  uint32_t key = GetCacheKey(filename_utf8.ToStringView(), type);
+  std::string file_path = filename_utf8.ToString();
+  // If the relative path is enabled, we try to use a relative path
+  // from the compile cache directory to the file path
+  if (use_relative_ && IsAbsoluteFilePath(file_path)) {
+    // Normalise the paths to ensure they are consistent.
+    std::string normalised_file_path = NormalisePath(file_path);
+    std::string normalised_cache_dir =
```

Comment:
This is something we can just compute once and store in the CompileCacheHandler instance. This is currently being repeatedly computed for every path match even though it's the same every time.